src/components/Editor/CodeEditor.tsx

```diff
@@ -0,0 +1,209 @@
+
+  const connectClient = (clientId: string) => {
+    const existedClient = peerClients[clientId];
+
+    let color: string;
+    if (existedClient && existedClient.status === ConnectionStatus.Disconnected) {
+      color = existedClient.color;
+    } else {
+      color = randomColor();
+    }
+
+    const newClientCursor = ClientCursor.of(clientId, color);
+    otherClientsCursor.current.set(clientId, newClientCursor);
+    dispatch(AddPeer(clientId, color));
+  };
+
+  // Attach document
+  useEffect(() => {
+    dispatch(loadDocAction(true));
+    dispatch(AttachDocAction(props.docKey));
+  }, [props.docKey, dispatch]);
+
+  // Subscribe other client
+  useEffect(() => {
+    if (!client || !doc) {
+      return;
+    }
+
+    const disconnectClient = (clientId: string) => {
+      if (otherClientsCursor.current.has(clientId)) {
+        otherClientsCursor.current.get(clientId)!.removeCursor();
+        otherClientsCursor.current.delete(clientId);
+      }
+      dispatch(DisconnectPeer(clientId));
+    };
+
+    const unsubscribe = client.subscribe((event: any) => {
+      if (event.name === 'documents-watching-peer-changed') {
+        const newPeerClientsId: string[] = event.value[doc.getKey().toIDString()];
+        const setNewPeerClientsId = new Set(newPeerClientsId);
+
+        for (const clientId of Object.keys(peerClients)) {
+          if (setNewPeerClientsId.has(clientId) && peerClients[clientId].status === ConnectionStatus.Connected) {
+            continue;
+          }
+          disconnectClient(clientId);
+        }
+      }
+    });
+
+    return () => {
+      unsubscribe();
+    };
+  }, [client, doc, peerClients, dispatch]);
+
+  if (loading) {
+    return (
+      <Box height="100%">
+        <CircularProgress color="secondary" />
+      </Box>
+    );
+  }
+
+  if (errorMessage || client === null || doc === null) {
+    return (
+      <div className={classes.root}>
+        <Alert severity="error">{errorMessage || 'fail to attach document'}</Alert>
+      </div>
+    );
+  }
+
+  return (
+    <CodeMirror
+      options={{ mode: 'xml', theme: 'monokai', lineNumbers: true }}
+      editorDidMount={(editor) => {
+        const updateCursor = (clientId: string, pos: number) => {
+          const clientCursor = otherClientsCursor.current.get(clientId);
+          clientCursor!.updateCursor(editor, pos);
+        };
+
+        const updateLine = (clientId: string, fromIdx: number, toIdx: number) => {
+          const clientCursor = otherClientsCursor.current.get(clientId);
+          clientCursor!.updateLine(editor, fromIdx, toIdx);
+        };
+
+        doc?.subscribe((event: any) => {
+          if (event.name === 'remote-change') {
+            event.value.forEach((change: any) => {
+              const actor = change.id.actor;
+              if (actor !== client.getID()) {
+                if (!otherClientsCursor.current.has(actor)) {
+                  connectClient(actor);
+                  updateCursor(actor, 0);
```

Comment:
If we put the type definition mentioned above, the second parameter is not a numeric type, so an error occurs in this code.

Reply:
@hackerwins correct it! :) If I don't use type,, I would not know that.. 😵  thank you @dc7303 👍